appliance/postgresql2/state/state.go

```diff
@@ -0,0 +1,1019 @@
+			continue
+		case e := <-postgresCh:
+			p.handlePgInit(e)
+			continue
+		case <-p.evalStateCh:
+			p.evalClusterState()
+			continue
+		case <-p.applyConfCh:
+			p.pgApplyConfig()
+			continue
+		default:
+			// There were no receivable channels, fallthrough to the select that
+			// includes the work done channel
+		}
+		select {
+		case e := <-discoverdCh:
+			p.handleDiscoverdEvent(e)
+		case e := <-postgresCh:
+			p.handlePgInit(e)
+		case <-p.evalStateCh:
+			p.evalClusterState()
+		case <-p.applyConfCh:
+			p.pgApplyConfig()
+		case <-p.workDoneCh:
+			// There is no work to do, we are now at rest
+			p.rest()
+		case <-p.stopCh:
+			return
+		}
+	}
+}
+
+func (p *Peer) Close() error {
+	close(p.stopCh)
+	return nil
+}
+
+type PeerInfo struct {
+	ID             string                `json:"id"`
+	Role           Role                  `json:"role"`
+	PgRetryPending *time.Time            `json:"pg_retry_pending,omitempty"`
+	Updating       bool                  `json:"updating,omitempty"`
+	State          *State                `json:"state"`
+	Peers          []*discoverd.Instance `json:"peers"`
+}
+
+func (p *Peer) Info() *PeerInfo {
+	return &PeerInfo{
+		ID:             p.self.ID,
+		Role:           p.role,
+		PgRetryPending: p.pgRetryPending,
+		State:          p.clusterState,
+		Peers:          p.clusterPeers,
+	}
+}
+
+func (p *Peer) handlePgInit(e PostgresEvent) {
+	p.log.Info("postgres init", "at", "init_pg", "online", e.Online, "setup", e.Setup)
+	if p.pgOnline != nil {
+		panic("received postgres init event after already initialized")
+	}
+
+	p.pgOnline = &e.Online
+	p.pgSetup = e.Setup
+
+	if p.clusterPeers != nil {
+		p.evalClusterState()
+	}
+}
+
+func (p *Peer) handleDiscoverdEvent(e *DiscoverdEvent) {
+	switch e.Kind {
+	case DiscoverdEventInit:
+		p.handleDiscoverdInit(e)
+	case DiscoverdEventState:
+		p.handleDiscoverdState(e)
+	case DiscoverdEventPeers:
+		p.handleDiscoverdPeers(e)
+	}
+}
+
+func (p *Peer) handleDiscoverdInit(e *DiscoverdEvent) {
+	p.log.Info("discoverd init", "at", "init_discoverd", "peers", len(e.Peers), "state", e.State.State != nil)
+	if p.clusterState != nil {
+		panic("received discoverd init after already initialized")
+	}
+	p.clusterPeers = e.Peers
+	p.decodeState(e)
+	if p.pgOnline != nil {
+		p.evalClusterState()
+	}
+}
+
+func (p *Peer) handleDiscoverdPeers(e *DiscoverdEvent) {
+	p.log.Info("discoverd peers", "at", "discoverd_peers", "peers", len(e.Peers))
+	if p.clusterPeers == nil {
+		panic("received discoverd peers before init")
+	}
+	p.clusterPeers = e.Peers
+	p.evalClusterState()
+}
+
+func (p *Peer) handleDiscoverdState(e *DiscoverdEvent) {
+	log := p.log.New("fn", "handleDiscoverdState", "index", e.State.Index)
+	log.Info("got discoverd state", "state", e.State.State != nil)
+	if p.clusterPeers == nil {
+		panic("received discoverd state before init")
+	}
+	if p.decodeState(e) {
+		p.evalClusterState()
+	} else {
+		log.Info("already have this state", "at", "skip_eval")
```

Comment:
Comment added to discuss: https://github.com/flynn/flynn/issues/34#issuecomment-75448708